src/modules/rlm_smtp/rlm_smtp.c

```diff
@@ -0,0 +1,614 @@
+	} else {
+		RDEBUG2("server certificate result not found");
+	}
+
+	if (randle->result != CURLE_OK) {
+		talloc_free(randle);
+		return RLM_MODULE_REJECT;
+	}
+
+	if (tls->extract_cert_attrs) fr_curl_response_certinfo(request, randle);
+
+	talloc_free(randle);
+	return RLM_MODULE_OK;
+}
+
+/*
+ *	Checks that there is a User-Name and User-Password field in the request
+ *	As well as all of the required SMTP elements
+ *	Sets the: username, password
+ *		website URI
+ *		timeout information
+ *		TLS information
+ *		Sender and recipient information
+ *		Email header and body
+ *		File attachments
+ *
+ *	Then it queues the request and yeilds until a response is given
+ *	When it responds, mod_authorize_resume is called.
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authorize(module_ctx_t const *mctx, REQUEST *request)
+{
+	rlm_smtp_t const		*inst = talloc_get_type_abort_const(mctx->instance, rlm_smtp_t);
+	rlm_smtp_thread_t       	*t = talloc_get_type_abort(mctx->thread, rlm_smtp_thread_t);
+	fr_curl_io_request_t     	*randle;
+	fr_mail_ctx			*mail_ctx;
+	curl_mime			*mime;
+	curl_mime			*mime_body;
+	curl_mimepart			*part;
+	int 				body_elements;
+
+	VALUE_PAIR const 		*vp, *smtp_header, *smtp_body, *sender_email, *username, *password;
+
+	if (fr_pair_find_by_da(request->control, attr_auth_type, TAG_ANY) != NULL) {
+		RDEBUG3("Auth-Type is already set.  Not setting 'Auth-Type := %s'", inst->name);
+		return RLM_MODULE_NOOP;
+	}
+	/* Elements provided by the request */
+	username = fr_pair_find_by_da(request->packet->vps, attr_user_name, TAG_ANY);
+	password = fr_pair_find_by_da(request->packet->vps, attr_user_password, TAG_ANY);
+	smtp_header = fr_pair_find_by_da(request->packet->vps, attr_smtp_header, TAG_ANY);
+	smtp_body = fr_pair_find_by_da(request->packet->vps, attr_smtp_body, TAG_ANY);
+	sender_email = fr_pair_find_by_da(request->packet->vps, attr_smtp_sender_email, TAG_ANY);
+
+	/* Make sure all of the essential email components are present and possible*/
+	if(!smtp_header) {
+		RDEBUG2("Attribute \"smtp-header\" is required for smtp");
+		return RLM_MODULE_INVALID;
+	}
+	if (smtp_header->vp_length == 0) {
+		RDEBUG2("\"smtp_header\" must not be empty");
+		return RLM_MODULE_INVALID;
+	}
+	if(!smtp_body) {
+		RDEBUG2("Attribute \"smtp-body\" is required for smtp");
+		return RLM_MODULE_INVALID;
+	}
+
+	/* allocate the handle and set the curl options */
+	randle = fr_curl_io_request_alloc(request);
+	if (!randle){
+		RDEBUG2("A handle could not be allocated for the request");
+		return RLM_MODULE_FAIL;
+	}
+
+	/* Initialize the mail_ctx to perform the email */
+	mail_ctx = talloc_zero(randle, fr_mail_ctx);
+	mail_ctx->request 		= request;
+	mail_ctx->randle 		= randle;
+	mail_ctx->time_status 		= UNLOADED; /* prevent duplication if there is a user supplied a DATE: element */
+	mail_ctx->time 			= fr_time(); /* time the request was received. Used to set DATE: if none is supplied */
```

Comment:
Minor nit with this.  Generally when you're allocating and immediately initialising a struct, don't use the zero or clearing allocation function, just initialise a structure with a compound literal. i.e.

``
*mail_ctx = (fr_mail_ctx){
	.request = request,
	.randle = randle,
	.time_status = UNLOADED,
	.time = fr_time()
}
```

Elements you don't explicitly specify will be initialised to zero, and you save writing over the struct twice (once for zeroing it out, and once for filling it).